File: 2024/6/6-1.cpp
#include <iostream>
#include <ostream>
#include <set>
#include <vector>

struct Coord {
  int x;
  int y;

  bool operator<(const Coord &other) const {
    if (x != other.x) {
      return x < other.x;
    }
    return y < other.y;
  }
};

struct Velocity {
  int x;
  int y;

  bool operator<(const Velocity &other) const {
    if (x != other.x) {
      return x < other.x;
    }
    return y < other.y;
  }
};

class Board {
public:
  std::set<Coord> obstacles;
  int max_x;
  int max_y;

  Board(std::vector<std::string> &lines) {
    max_x = lines.size();
    max_y = lines[0].size();
    for (int row = 0; row < lines.size(); ++row) {
      for (int column = 0; column < lines[row].size(); ++column) {
        if (lines[row][column] == '#') {
          obstacles.emplace(Coord{column, row});
        }
      }
    }
  }
};

class Guard {
  Coord position;
  Velocity velocity;
  Board board;

public:
  std::set<Coord> visited_positions;
  Guard(std::vector<std::string> &lines, Board &board) : board(board) {
    for (int row = 0; row < lines.size(); ++row) {
      for (int column = 0; column < lines[row].size(); ++column) {
        char current_char = lines[row][column];
        if (current_char == '^') {
          velocity = Velocity{0, -1};
          position = {column, row};
        } else if (current_char == '>') {
          velocity = Velocity{1, 0};
          position = {column, row};
        } else if (current_char == '<') {
          velocity = Velocity{-1, 0};
          position = {column, row};
        } else if (current_char == 'v') {
          velocity = Velocity{0, 1};
          position = {column, row};
        }
      }
    }
  }

  bool inBounds() {
    return (position.x > 0 && position.y > 0 && position.x < board.max_x &&
            position.y < board.max_y);
  }

  bool collision() {
    return (board.obstacles.find(
                Coord{position.x + velocity.x, position.y + velocity.y}) !=
            board.obstacles.end());
  }

  void rotate() {
    int old_x = velocity.x;
    velocity.x = -velocity.y;
    velocity.y = old_x;
  }

  void move() {
    while (collision()) {
      rotate();
    }
    position.x += velocity.x;
    position.y += velocity.y;
    visited_positions.emplace(position);
  }
};

int main() {
  std::string line;
  std::vector<std::string> lines;
  int output = 0;
  while (std::getline(std::cin, line)) {
    lines.emplace_back(line);
  };
  Board board = Board(lines);
  Guard guard = Guard(lines, board);
  while (guard.inBounds()) {
    guard.move();
  };

  std::cout << "visited: " << guard.visited_positions.size() << std::endl;
  return 0;
}

File: 2024/6/6-1.py
from dataclasses import dataclass
import sys
import time


@dataclass
class Coord:
    x: int
    y: int

    def __hash__(self):
        return hash((self.x, self.y))


@dataclass
class Velocity:
    x: int
    y: int


class Board:
    max_x: int
    max_y: int
    obstacles: set[Coord]

    def __init__(self, lines: list[str]):
        self.obstacles = set()
        self.max_x = len(lines)
        self.max_y = len(lines[0])

        for row, line in enumerate(lines):
            for column, char in enumerate(line):
                if char == "#":
                    self.obstacles.add(Coord(x=column, y=row))

    def draw(self, guard: "Guard"):
        sys.stdout.write("\033[H")  # Move cursor to the top-left
        sys.stdout.write("\033[J")  # Clear the screen below the cursor
        for row in range(0, self.max_x):
            for column in range(0, self.max_y):
                if Coord(x=column, y=row) in self.obstacles:
                    print("#", end="")
                elif Coord(x=column, y=row) == guard.position:
                    print(guard.char, end="")
                else:
                    print(".", end="")
            print("\n")
        print(guard.position)
        print(guard.velocity)
        time.sleep(0.2)
        sys.stdout.flush()


class Guard:
    position: Coord
    velocity: Velocity
    board: Board
    visited_positions: set[Coord] = set()

    def __init__(self, lines: list[str], board: Board):
        self.board = board
        for row, line in enumerate(lines):
            for column, char in enumerate(line):
                position = Coord(x=column, y=row)
                if char == "^":
                    self.position = position
                    self.velocity = Velocity(x=0, y=-1)
                if char == ">":
                    self.position = position
                    self.velocity = Velocity(x=1, y=0)
                if char == "<":
                    self.position = position
                    self.velocity = Velocity(x=-1, y=0)
                if char == "v":
                    self.position = position
                    self.velocity = Velocity(x=0, y=1)

    @property
    def char(self):
        if self.velocity.x < 0:
            return "<"
        if self.velocity.x > 0:
            return ">"
        if self.velocity.y > 0:
            return "v"
        if self.velocity.y < 0:
            return "^"

    def move(self):
        if self.collision():
            self.rotate()
        self.position.x += self.velocity.x
        self.position.y += self.velocity.y
        self.visited_positions.add(self.position)
        # self.board.draw(self)

    def rotate(self):
        self.velocity.x, self.velocity.y = -self.velocity.y, self.velocity.x

    def collision(self):
        return (
            Coord(
                x=(self.position.x + self.velocity.x),
                y=self.position.y + self.velocity.y,
            )
            in self.board.obstacles
        )

    def in_bounds(self):
        return (
            self.position.x > 0
            and self.position.y > 0
            and self.position.x < self.board.max_x
            and self.position.y < self.board.max_y
        )


with open("input.txt", "r") as f:
    lines = f.read().splitlines()

guard = Guard(lines, Board(lines))
while guard.in_bounds():
    guard.move()

print(len(guard.visited_positions))
